### ISSUE TITLE
Code review

### ISSUE THREAD

**Issue by Gondolav** [favo-sdp/Favo#209]
# Code review

This is a code review for the code of the whole project on master branch at commit https://github.com/favo-sdp/Favo/tree/fde932d4406b5853cdfc3bdc48a35f108ee18aa6.

Of course, we don't expect you to change all of your code (some parts of the review can be subjective or arguable anyway), what’s important is that you understand why better alternatives may exist.

Please ask us if you have any question about the code review, whether it is by writing a comment below or by sending us an email or meeting us.

**Important: if you try to fix these, don't break your app. We prefer that the app works even if you don't manage to fix anything. It may be too hard to rework many of the issues below, as it would require to rewrite the whole architecture.**

## General remarks

Most of the general remarks are listed here; only very file-specific remarks are listed under each file section.

- Prefer generic types over raw types. You have many warnings generated by Android Studio about that, you should fix them.
- You can (and should) run the Android linter to spot errors and warnings (*Analyze -> Inspect Code...*).
- You should always override `hashCode()` whenever you override `equals()`. You can generate both automatically with Android Studio by going to *Code > Generate...* in the menu.
- Access modifiers of fields and methods should always be as restrictive as possible.
- Always add JavaDoc for public and protected functions, especially for interfaces. Move comments inside interfaces if you defined them in the classes implementing the interfaces.
- Commenting code to remove it (temporarily or not) is “ok” but only on a local branch. On master, it should not occur, PR reviews should point these pieces of code out and they should be removed before merging. They make the code hard to read.
- Constants should be defined as `static final`.
- You should avoid using magic numbers in your code. Define constants instead.
- You should always use constants for intents' extras' (or bundles) keys, and either define them public in one of the classes using them, or define a file containing only constants.
- You put sometimes string literals directly into the code. You should always use Android resources for strings that will appear on the UI. It is useful especially for translation but also just for “having everything at the same place”. See Android Linter for more. That is for xml and java files.

#### SignInActivity.java

- `getCurrentUserUtil()` is not very useful as it only calls a static method.
- You use raw types in `handleSignInResponse()` and `postNewUser()` for CompletableFuture. Use instead generic types.
- You don't need the local variable `postNewUser` in `postNewUser()`.

#### ChatModel.java

- You can simply use `Objects.hash()` in `hashCode()`. Furthermore, you can generate both `equals()` and `hashCode()` with Android Studio (Code -> Generate...).

#### ChatPage.java

- `setupToolBar()` can be private.

#### DatabaseWrapper.java

- Some functions can be package-private.

#### ICollectionWrapper.java

- You use many raw types. Use instead generic types.
- You put `T extends Document` in the class, you should put it in the interface, as the `ICollectionWrapper` deals only with documents.
- The choice of name is a bit too generic, as it wraps a collection of `Documents` is should be clear in the name as well.

#### Favor.java

- You can define `equals` based on the attributes you want. It is not necessary to compare all attributes. So you can redefine `equals` instead of having `contentEquals`.
- The use of `clone` is not encouraged, you should prefer a copy constructor, that takes a given `Favor` and creates a new one with the same values. `clone` method could be tricky and returns an `Object` and that forces you to cast it, which is not pretty.

#### FavorStatus.java

- `archivedStates` can be static final (and should if it is constant).

#### FavorUtil.java

- You should not have `NotImplementedException()`. If you don't need a method, just remove it on master. It adds possibilities to add bugs to your code.
- `TAG` and `COLLECTION_NAME` are never used.
#### FavoLocation.java

- `EARTH_RADIUS` should be static final.
- `equals` looks weird, you should generate it (now you know you can :) )

#### GpsTracker.java

- [x] There are a few magic numbers that can be converted into constants.
- [x] Does it need to extend `FragmentActivity`? Also, it is defined as a service in the manifest, even though it extends an activity.

#### FirebaseMessagingService.java

- You should implement `onNewToken()`.
- Same comment as before about commented code.

#### IUserUtil.java

- You use many raw types. Use instead generic types.

#### User.java

- You can restrict the access modifier of many fields and methods.

#### UserUtil.java

- `TAG` is never used.
- Local variable `user` in `changeActiveFavorCount()` is redundant. You can use `object` or take `user` directly as lambda-parameter.

#### BitmapConversionUtil.java

- You can create a constant for the quality parameter in `compress()` inside `bitmapToJpegInputStream()`.
- In `bitmapToJpegInputStream`, you should inline the variable `bais`.

#### DependencyFactory.java

- Two fields are not used.
- You should never use flags like `testMode` in production code. Here, you could simply initialize by default the fields to the real instances, and call setters on them for dependency injection in tests. By proceeding like this, the production code "has no idea" of tests at all.

#### PictureUtil.java
- To complete the comment above about `testMode`: this class would have to provide a setter to enable dependency injection (without `testMode`).

#### TaskToFutureAdapter.java

- Change this class such that you don't use `isTestMode()` too.

##### MapPage.java

- [x] Replace magic numbers and string literals with constants.

##### UserAccountPage.java

- `currentUser` is never accessed.
- In `displayUserData`, the string used in the message should be in Android Resources, as stated in _General remarks_.
- Same in `deleteAccountClicked`.

###### FavorDetailView.java

- `locationAccessBtn` can be a local variable.
- Use constants for bundles' keys, as suggested in *General remarks*.

###### FavorRequestView.java

- `TAG` should be static final.
- `locationAccessBtn` can be a local variable.
- In `setupButtons()` there is an empty if statement in the click listener of `cancelFavorBtn`.
- You can use `==` on enums (in `setupButtons()`, for example). Also, in this case you can (and should) use a `switch`.
- In `isCameraAvailable()` you use a deprecated method (`getNumberOfCameras()`). You should avoid it.

###### FavorPage.java

- Replace magic number with constants.

#### IFavorViewModel.java

- Same remarks as before about raw types.
- Again, constants have to be `static final`.
- Some fields and methods can be private.

#### MainActivity.java

- [x] `cnt` is never used.
- [x] Use constants for Intent extras tags.
- [x] Use Android Resources for strings displayed in the UI (as stated in _General remarks_).

## Tests

- Testing code should be as readable and clean as production code. So same applies about commented pieces of code.
- All tests should meaningfully assert conditions or check useful properties (and not just click on buttons). Tests should test more than just "it didn't crash".

# Conclusion
Overall your code is good. The only real conception flow is the use of `testMode`. Another problem are your tests, which are not always meaningful. You have some small issues or adjustments that you could do but your codebase is rather good.

**Comment by huajian1069** [comment]
safe code clean in MapPage, GpsTracker file done, in my branch refactor/request-accept-favor-view

**Comment by mikanikos** [comment]
@Gondolav @samuelchassot 

Thanks a lot for providing the feedback! I just have one question about the testMode issue: we indeed tried the classic approach we studied in Sweng (i.e. initialize the static fields to the real instances and then using dependency injection with setters in tests, similarly to this exercise https://github.com/sweng-epfl/public/blob/master/exercises/testing/solutions/adapt-the-code.md#dependency-injection-on-android) but I think this doesn't always work for every dependency.

For example, let's take the currentFirebaseUser dependency which should be initialized at the beginning to FirebaseAuth.getInstance().getCurrentUser() (the real instance). If we initialize the dependency this way, we might not get the change of this variable during the execution. In fact, the real instance can be updated: it is null when no user is logged but it automatically becomes non null when a user signs in thanks to the FirebaseUI Auth library. So, if we just rely on initializing this variable at the beginning, when calling getCurrentFirebaseUser we could get always the same initial version of the real instance in production despite the change (in this specific case, a visible problem would be the impossibility to sign in or sign out while using the app).

I hope the problem is clear, we can also better discuss this during the next call if there's enough time. Probably there's a trivial solution to this or we're not considering something important but so far I think none of us couldn't find a better way to integrate dependency injection in tests as well as keeping the app work correctly in this specific case.

Thanks in advance for your help! 👍
